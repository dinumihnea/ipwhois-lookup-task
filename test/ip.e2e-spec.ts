import * as request from 'supertest';

export const APP_URL = `http://localhost:${process.env.APP_PORT}`;

describe('IpController (e2e)', () => {
  describe('/ips/ (GET)', () => {
    it.each([
      {
        query: '',
        errors: {
          ip: 'ip must be an ip address, ip should not be empty',
        },
      },
      {
        query: 'ip=',
        errors: {
          ip: 'ip must be an ip address, ip should not be empty',
        },
      },
      {
        query: 'ip=INVALIUD',
        errors: {
          ip: 'ip must be an ip address',
        },
      },
    ])('422 - should reject when $query', ({ query, errors }) => {
      return request(APP_URL)
        .get('/ips')
        .query(query)
        .expect(422)
        .expect((res) => {
          expect(res.body).toEqual({
            status: 422,
            errors,
          });
        });
    });

    it('200 - should return details about given ip', () => {
      return request(APP_URL)
        .get('/ips')
        .query({ ip: '8.8.4.4' })
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot({
            timezone: {
              current_time: expect.any(String),
            },
          });
        });
    });
  });

  describe('/ips/ (DELETE)', () => {
    it.each([
      {
        query: '',
        errors: {
          ip: 'ip must be an ip address, ip should not be empty',
        },
      },
      {
        query: 'ip=',
        errors: {
          ip: 'ip must be an ip address, ip should not be empty',
        },
      },
      {
        query: 'ip=INVALIUD',
        errors: {
          ip: 'ip must be an ip address',
        },
      },
    ])('422 - should reject when $query', ({ query, errors }) => {
      return request(APP_URL)
        .delete('/ips')
        .query(query)
        .expect(422)
        .expect((res) => {
          expect(res.body).toEqual({
            status: 422,
            errors,
          });
        });
    });

    it('204 - should delete details when previously stored', async () => {
      const { ip } = await request(APP_URL)
        .get('/ips')
        .query({ ip: '8.8.4.4' })
        .expect(200)
        .then(({ body }) => body);

      await request(APP_URL).delete('/ips').query({ ip }).expect(204);
    });
  });
});
