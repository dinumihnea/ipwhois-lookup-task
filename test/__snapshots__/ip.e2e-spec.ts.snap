// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IpController (e2e) ips/ (GET) 200 - should return details about given ip 1`] = `
{
  "borders": "CA,MX",
  "calling_code": "1",
  "capital": "Washington D.C.",
  "city": "Mountain View",
  "connection": {
    "asn": 15169,
    "domain": "google.com",
    "isp": "Google LLC",
    "org": "Google LLC",
  },
  "continent": "North America",
  "continent_code": "NA",
  "country": "United States",
  "country_code": "US",
  "flag": {
    "emoji": "ðŸ‡ºðŸ‡¸",
    "emoji_unicode": "U+1F1FA U+1F1F8",
    "img": "https://cdn.ipwhois.io/flags/us.svg",
  },
  "ip": "8.8.4.4",
  "is_eu": false,
  "latitude": 37.3860517,
  "longitude": -122.0838511,
  "postal": "94039",
  "region": "California",
  "region_code": "CA",
  "success": true,
  "timezone": {
    "abbr": "PST",
    "current_time": Any<String>,
    "id": "America/Los_Angeles",
    "is_dst": false,
    "offset": -28800,
    "utc": "-08:00",
  },
  "type": "IPv4",
}
`;
